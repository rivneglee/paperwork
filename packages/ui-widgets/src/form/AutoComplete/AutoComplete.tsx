import React, { FunctionComponent, ReactElement } from 'react';
import ReactSelect from 'react-select';
import AsyncSelect from 'react-select/async';

import { FieldGroup } from '../FieldGroup';

export interface SelectOption {
  value: string | number;
  label: string;
}

type SelectedValue = string | string[] | number | number[] | undefined;

interface Props {
  options: SelectOption[];
  selectedValue?: SelectedValue;
  label?: ReactElement | string;
  disabled?: boolean;
  isRequired?: boolean;
  labelAccessory?: ReactElement;
  isMultipleSelect?: boolean;
  onChange?: (value: SelectedValue) => void;
  size?: 'xs' | 's' | 'm' | 'l' | 'xl';
  labelPlacement?: 'left' | 'top';
  loadOptions?: (inputValue: string) => Promise<SelectOption[]>;
  noOptionsMessage?: string;
}

const Select: FunctionComponent<Props> = ({
  label,
  isRequired,
  labelAccessory,
  options = [],
  selectedValue,
  isMultipleSelect = false,
  disabled = false,
  onChange,
  size,
  labelPlacement = 'left',
  loadOptions,
  noOptionsMessage,
}) => {
  let selection = null;
  if (selectedValue instanceof Array) {
    selection = (selectedValue as []).map((
      s => options.find(o => o.value === s)
    ));
  } else {
    selection = options.find(o => o.value === selectedValue);
  }

  const onValueChange = (selection: SelectOption | SelectOption[]) => {
    let currentValue;
    if (selection && (selection as SelectOption).value) {
      currentValue = (selection as SelectOption).value;
    } else if (selection instanceof Array) {
      currentValue = selection.map(s => s.value) as string[];
    }

    if (onChange) {
      onChange(currentValue);
    }
  };

  const getNoOptionsMessage = () => {
    const defaultMessage = loadOptions ? 'Type in keyword to search options' : 'No options';
    return noOptionsMessage || defaultMessage;
  };

  const view = loadOptions ? (
    <AsyncSelect
      isMulti={isMultipleSelect}
      options={options}
      className="pw-select"
      classNamePrefix="pw-select"
      onChange={onValueChange}
      defaultValue={selection}
      isDisabled={disabled}
      loadOptions={loadOptions}
      noOptionsMessage={getNoOptionsMessage}
    />
  ) : (
    <ReactSelect
      isMulti={isMultipleSelect}
      options={options}
      className="pw-select"
      classNamePrefix="pw-select"
      onChange={onValueChange}
      defaultValue={selection}
      isDisabled={disabled}
      noOptionsMessage={getNoOptionsMessage}
    />
  );

  return (
    <FieldGroup
      label={label}
      isRequired={isRequired}
      labelAccessory={labelAccessory}
      size={size}
      labelPlacement={labelPlacement}
    >
      {view}
    </FieldGroup>
  );
};

export default Select;
