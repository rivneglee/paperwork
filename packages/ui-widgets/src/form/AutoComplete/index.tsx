export { default as AutoComplete } from './AutoComplete';
export * from './AutoComplete';
