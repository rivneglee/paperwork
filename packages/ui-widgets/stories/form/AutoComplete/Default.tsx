import React, { useState } from 'react';

import { AutoComplete } from '../../../src';

export default () => {
  const [value, setValue] = useState({ id: 'g', displayName: 'Green' });
  const colours = [
    { id: 'r', displayName: 'Red' },
    { id: 'g', displayName: 'Green' },
    { id: 'b', displayName: 'Blue' },
  ];

  const loadOptions = (inputValue: string): Promise<any[]> => (
    new Promise(resolve => (
        setTimeout(() =>
          resolve(colours.filter(
            c => c.displayName.indexOf(inputValue) !== -1)), 1000)
      ),
    )
  );

  return (
    <div style={{ background: '#ebeef1', padding: '2.4rem' }}>
      <AutoComplete
        loadOptions={loadOptions}
        label="Auto complete"
        value={value}
        size="m"
        isMultipleSelect={true}
        onChange={(selection: any) => setValue(selection)}
        getOptionLabel={option => option.displayName}
        getOptionValue={option => option.id}
      />
    </div>
  );
};
