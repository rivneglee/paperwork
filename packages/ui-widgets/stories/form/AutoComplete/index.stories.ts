import { storiesOf } from '@storybook/react';

import Default from './Default';

storiesOf('AutoComplete', module)
  .add('default', Default);
