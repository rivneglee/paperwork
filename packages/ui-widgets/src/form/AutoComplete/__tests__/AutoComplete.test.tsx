import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import AutoComplete from '../AutoComplete';

describe('AutoComplete', () => {
  let wrapper: ShallowWrapper;

  const colours = [
    { id: 'r', displayName: 'Red' },
    { id: 'g', displayName: 'Green' },
    { id: 'b', displayName: 'Blue' },
  ];

  const loadOptions = (inputValue: string): Promise<any[]> => (
    new Promise(resolve => (
        setTimeout(() =>
          resolve(colours.filter(
            c => c.displayName.indexOf(inputValue) !== -1)), 500)
      ),
    )
  );

  beforeAll(() => {
    wrapper = shallow(
      <AutoComplete
        loadOptions={loadOptions}
        value={{ id: 'r', displayName: 'Red' }}
        getOptionLabel={option => option.displayName}
        getOptionValue={option => option.id}
      />,
    );
  });

  it('should render properly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
