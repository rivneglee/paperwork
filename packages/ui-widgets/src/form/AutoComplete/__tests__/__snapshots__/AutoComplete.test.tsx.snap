// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete should render properly 1`] = `
<FieldGroup
  labelPlacement="left"
>
  <Async
    cacheOptions={false}
    className="pw-select"
    classNamePrefix="pw-select"
    components={
      Object {
        "DropdownIndicator": [Function],
      }
    }
    defaultOptions={false}
    defaultValue={
      Object {
        "displayName": "Red",
        "id": "r",
      }
    }
    filterOption={null}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    isDisabled={false}
    isLoading={false}
    isMulti={false}
    loadOptions={[Function]}
    noOptionsMessage={[Function]}
    options={
      Array [
        Object {
          "displayName": "Red",
          "id": "r",
        },
      ]
    }
    placeholder=""
  />
</FieldGroup>
`;
